### PR TITLE
"Password-ing"

### PR DESCRIPTION
**Pull request by Stokalas**
- Ability to recover forgotten password
- Ability to change password when logged in
- When password recovery code sent but password not changed reminder in index.
- Password recovery code is valid: 15min; until logged in/registered; until recovery canceled;
- Password recovery page should not be reachable when recovery is not in progress.
- Change password page should not be reachable when not logged in.

### CODE REVIEW

**Comment by Stokalas** [comment]
**Note to myself**: to add client side input validation before merging this pull request 🙂

File: src/Pages/logIn.cshtml.cs [hunk 1]
```diff
@@ -56,8 +56,8 @@ public async Task<IActionResult> OnPost(string returnUrl)
         public IActionResult OnPostForgotPassword()
         {
             if (!ModelState.IsValid) return Page();
-
-            Response.Cookies.Append("Email", Email);
+            var option = new CookieOptions() { Expires = DateTime.Now.AddMinutes(15) };
```
Comment (DeividasLT5) [inline src/Pages/logIn.cshtml.cs:82]: Just wondering, how does this expiration work? In a little more detail, please

File: src/Pages/Index.cshtml [hunk 2]
```diff
@@ -4,7 +4,16 @@
...
+@if(Request.Cookies.ContainsKey("recoveryCode"))
+{
+    <div class="alert alert-warning" role="alert">
+        Password recovery is in progress! <a asp-page="/forgotPassword" class="alert-link">Go to recovery</a> to finish it!.
+        <button type="button" class="close" data-dismiss="alert" aria-label="Close">
+            <span aria-hidden="true">&times;</span>
+        </button>
+    </div>
+   
+}
```
Comment (ciliakas) [inline src/Pages/Index.cshtml:16]: Not sure if this is supposed to work this way, but if you start password recovery, and then login without finishing it, this popup doesn't appear

File: src/Pages/logIn.cshtml.cs [hunk 3]
```diff
@@ -43,14 +52,25 @@ public async Task<IActionResult> OnPost(string returnUrl)
                 await HttpContext.SignInAsync(CookieAuthenticationDefaults.AuthenticationScheme,
                     new ClaimsPrincipal(claimsIdentity));
 
+                Response.Cookies.Delete("recoveryCode");
+                Response.Cookies.Delete("Email");
                 return Redirect(returnUrl ?? "/Index");
-                
-
             };
```
Comment (ciliakas) [inline src/Pages/logIn.cshtml.cs]: Super minor thing, but this semicolon is unnecessary :D

File: src/Pages/ForgotPassword.cshtml.cs [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        [BindProperty]
+        [DataType(DataType.Password)]
+        public string PasswordConfirm { get; set; }
+
+        [BindProperty]
+        public string Email { get; set; }
+
+        public string ErrorMessage = "";
+
+        public bool CodeSent = false;
```
Comment (ciliakas) [inline src/Pages/ForgotPassword.cshtml.cs]: Resharper is telling me it's unnecessary to set CodeSent default value.

**Review by ciliakas — CHANGES_REQUESTED** [review]
Neato, good feature to have. Code looks good, left a few minor questions in the comments.
Few errors I've managed to spot though:
- The red error labels for the change password form are hard to read and kind of hurt the eyes
- If I start the password recovery, then login in another tab, and then finish password recovery, I get a blank screen and the password doesn't change. Granted, this is really unlikely to happen, but an error message here or something would be good to have.
Screenshots:
got this by calling recovery then changing then finishing recovery
text is hard to read

File: src/Pages/Index.cshtml (hunk 2 above)
Comment (Stokalas) [reply src/Pages/Index.cshtml:16]: Yup, it shouldn't since it resets recover if logged in.

File: src/Pages/ForgotPassword.cshtml.cs (hunk 4 above)
Comment (Stokalas) [reply src/Pages/ForgotPassword.cshtml.cs]: I guess bools by default are assigned to false.

File: src/Pages/logIn.cshtml.cs (hunk 1 above)
Comment (Stokalas) [reply src/Pages/logIn.cshtml.cs:82]: It just an option you can add, how the cookies are handled in the deep back I don't really know :D
I guess it just has a date, from which it is deleted.

**Review by ciliakas — APPROVED** [review]
All issues fixed

**Pull request merged**